### DISCUSSION TITLE
Video Streaming

### DISCUSSION THREAD

**Discussion by OmegaRogue** [diamondburned/gotk4#100] [Help]
Is there a way to display a video from an h264 byte stream on a video widget or does this work together with https://github.com/tinyzimmer/go-gst ?

**Comment by diamondburned** [comment]
Yes, you'd just use `GtkVideo` which internally uses either Gst or FFmpeg. `GtkMediaStream` works too, albeit they're both hard to plug an `io.Reader` into. `core/gioutil` might be very helpful here.

**Comment by OmegaRogue** [comment]
Thanks, ill try it out!

**Comment by OmegaRogue** [comment]
> Yes, you'd just use `GtkVideo` which internally uses either Gst or FFmpeg. `GtkMediaStream` works too, albeit they're both hard to plug an `io.Reader` into. `core/gioutil` might be very helpful here.

With `GtkVideo`, I just use a `GtkMediaStream` with `core/gioutil`, right? or is there another way to get a raw bytestream from memory into it?

**Comment by OmegaRogue** [comment]
So, i tried doing this
```go
pr, pw := io.Pipe()
bufpipe := bufio.NewReadWriter(bufio.NewReader(pr),bufio.NewWriter(pw))
video := gtk.NewVideoForMediaStream(gtk.NewMediaFileForInputStream(gioutil.NewInputStream(bufpipe)))
video.SetVExpand(true)
video.SetHExpand(true)
```
While in another part of my application bytes of the video are being written into the ReadWriter
```go
func (w *WeylusClient) Run() {
	// avoid racecondition
	time.Sleep(time.Second)
	for {
		select {
		case <-w.ctx.Done():
			log.Ctx(w.ctx).Err(errors.Wrap(w.ctx.Err(), "closed context")).Msg("closed context")
			return
		case msg := <-w.msgs:
			switch msg.Type {
			case websocket.MessageText:
				log.Ctx(w.ctx).Info().RawJSON("data", msg.Data).Msg("received data")
				for response, callbacks := range w.callbacks {
					if strings.Contains(string(msg.Data), string(response)) {
						for _, callback := range callbacks {
							callback(msg)
						}
					}
				}
			case websocket.MessageBinary:
				if _, err := w.BufPipe.Write(msg.Data); err != nil {
					log.Ctx(w.ctx).Err(err).Msg("error on write data")
				}
			}
		}
	}
}
```
And when running this, i get
```
**
ERROR:../gtk/modules/media/gtkgstmediafile.c:299:gtk_gst_media_file_open: code should not be reached
Bail out! ERROR:../gtk/modules/media/gtkgstmediafile.c:299:gtk_gst_media_file_open: code should not be reached
SIGABRT: abort
PC=0x7fe8b69a58ec m=0 sigcode=18446744073709551610
signal arrived during cgo execution

goroutine 1 [syscall, locked to thread]:
runtime.cgocall(0x1453b00, 0xc00020ef28)
        /home/omegarogue/go/go1.20.2/src/runtime/cgocall.go:157 +0x5c fp=0xc00020ef00 sp=0xc00020eec8 pc=0x56a6fc
github.com/diamondburned/gotk4/pkg/gtk/v4._Cfunc_gtk_media_file_new_for_input_stream(0x47a29a0)
        _cgo_gotypes.go:33137 +0x49 fp=0xc00020ef28 sp=0xc00020ef00 pc=0xbaf009
github.com/diamondburned/gotk4/pkg/gtk/v4.NewMediaFileForInputStream.func1(0xc0001e30a0?)
        /home/omegarogue/go/pkg/mod/github.com/diamondburned/gotk4/pkg@v0.0.5/gtk/v4/gtkmediafile.go:225 +0x3a fp=0xc00020ef60 sp=0xc00020ef28 pc=0xcda0da
github.com/diamondburned/gotk4/pkg/gtk/v4.NewMediaFileForInputStream({0x193c790?, 0xc000015d48})
        /home/omegarogue/go/pkg/mod/github.com/diamondburned/gotk4/pkg@v0.0.5/gtk/v4/gtkmediafile.go:225 +0x65 fp=0xc00020ef90 sp=0xc00020ef60 pc=0xcd9fe5
weylus-surface/cmd.activate(0x80?)
        /home/omegarogue/projects/weylus-surface/cmd/root.go:112 +0x454 fp=0xc00020f148 sp=0xc00020ef90 pc=0x13f94d4
weylus-surface/cmd.NewRootCmd.func1.1()
        /home/omegarogue/projects/weylus-surface/cmd/root.go:60 +0x1d fp=0xc00020f160 sp=0xc00020f148 pc=0x13fa43d
github.com/diamondburned/gotk4/pkg/gio/v2._gotk4_gio2_Application_ConnectActivate(0x56ace6?, 0xc0000061a0?)
        /home/omegarogue/go/pkg/mod/github.com/diamondburned/gotk4/pkg@v0.0.5/gio/v2/gio-2.0_export.go:714 +0x74 fp=0xc00020f1a8 sp=0xc00020f160 pc=0x86c2f4
_cgoexp_8c20477d56c3__gotk4_gio2_Application_ConnectActivate(0x7ffe6e31f3c0?)
        _cgo_gotypes.go:37543 +0x25 fp=0xc00020f1c8 sp=0xc00020f1a8 pc=0x928925
runtime.cgocallbackg1(0x928900, 0xc00020f378?, 0x0)
        /home/omegarogue/go/go1.20.2/src/runtime/cgocall.go:315 +0x2b1 fp=0xc00020f290 sp=0xc00020f1c8 pc=0x56abf1
runtime.cgocallbackg(0xc0000061a0?, 0x300000002?, 0xc0000061a0?)
        /home/omegarogue/go/go1.20.2/src/runtime/cgocall.go:234 +0x109 fp=0xc00020f320 sp=0xc00020f290 pc=0x56a8a9
runtime.cgocallbackg(0x928900, 0x7ffe6e31f750, 0x0)
        <autogenerated>:1 +0x2f fp=0xc00020f348 sp=0xc00020f320 pc=0x5d1d6f
runtime.cgocallback(0x56a725, 0x1416bd0, 0xc00020f3d8)
        /home/omegarogue/go/go1.20.2/src/runtime/asm_amd64.s:998 +0xb4 fp=0xc00020f370 sp=0xc00020f348 pc=0x5cf3f4
runtime.systemstack_switch()
        /home/omegarogue/go/go1.20.2/src/runtime/asm_amd64.s:463 fp=0xc00020f378 sp=0xc00020f370 pc=0x5cd400
runtime.cgocall(0x1416bd0, 0xc00020f3d8)
        /home/omegarogue/go/go1.20.2/src/runtime/cgocall.go:167 +0x85 fp=0xc00020f3b0 sp=0xc00020f378 pc=0x56a725
github.com/diamondburned/gotk4/pkg/gio/v2._Cfunc_g_application_run(0x461fa30, 0x1, 0x4621410)
        _cgo_gotypes.go:13957 +0x4c fp=0xc00020f3d8 sp=0xc00020f3b0 pc=0x7986ac
github.com/diamondburned/gotk4/pkg/gio/v2.(*Application).Run.func3(0xc0001e2cb0?, 0x27?, 0x4621410?)
        /home/omegarogue/go/pkg/mod/github.com/diamondburned/gotk4/pkg@v0.0.5/gio/v2/gapplication_2_28.go:1266 +0x71 fp=0xc00020f420 sp=0xc00020f3d8 pc=0x7eb9d1
github.com/diamondburned/gotk4/pkg/gio/v2.(*Application).Run(0xc0001bf6b0, {0xc000024240?, 0x1, 0x1})
        /home/omegarogue/go/pkg/mod/github.com/diamondburned/gotk4/pkg@v0.0.5/gio/v2/gapplication_2_28.go:1266 +0x1e5 fp=0xc00020f4e0 sp=0xc00020f420 pc=0x7eb8a5
weylus-surface/cmd.NewRootCmd.func1(0xc000004300?, {0x178e05c?, 0x0?, 0x0?})
        /home/omegarogue/projects/weylus-surface/cmd/root.go:61 +0xb4 fp=0xc00020f538 sp=0xc00020f4e0 pc=0x13fa3f4
github.com/spf13/cobra.(*Command).execute(0xc000004300, {0xc000024240, 0x0, 0x0})
        /home/omegarogue/go/pkg/mod/github.com/spf13/cobra@v1.6.1/command.go:920 +0x847 fp=0xc00020f670 sp=0xc00020f538 pc=0x11cc8a7
github.com/spf13/cobra.(*Command).ExecuteC(0xc000004300)
        /home/omegarogue/go/pkg/mod/github.com/spf13/cobra@v1.6.1/command.go:1044 +0x3bd fp=0xc00020f728 sp=0xc00020f670 pc=0x11cd13d
github.com/spf13/cobra.(*Command).Execute(...)
        /home/omegarogue/go/pkg/mod/github.com/spf13/cobra@v1.6.1/command.go:968
weylus-surface/cmd.Execute()
        /home/omegarogue/projects/weylus-surface/cmd/root.go:179 +0x25 fp=0xc00020f760 sp=0xc00020f728 pc=0x13f9f45
main.main()
        /home/omegarogue/projects/weylus-surface/main.go:50 +0xca5 fp=0xc00020ff80 sp=0xc00020f760 pc=0x13fd965
runtime.main()
        /home/omegarogue/go/go1.20.2/src/runtime/proc.go:250 +0x207 fp=0xc00020ffe0 sp=0xc00020ff80 pc=0x59fc87
runtime.goexit()
        /home/omegarogue/go/go1.20.2/src/runtime/asm_amd64.s:1598 +0x1 fp=0xc00020ffe8 sp=0xc00020ffe0 pc=0x5cf621

goroutine 2 [force gc (idle)]:
runtime.gopark(0x0?, 0x0?, 0x0?, 0x0?, 0x0?)
        /home/omegarogue/go/go1.20.2/src/runtime/proc.go:381 +0xd6 fp=0xc00005cfb0 sp=0xc00005cf90 pc=0x5a00b6
runtime.goparkunlock(...)
        /home/omegarogue/go/go1.20.2/src/runtime/proc.go:387
runtime.forcegchelper()
        /home/omegarogue/go/go1.20.2/src/runtime/proc.go:305 +0xb0 fp=0xc00005cfe0 sp=0xc00005cfb0 pc=0x59fef0
runtime.goexit()
        /home/omegarogue/go/go1.20.2/src/runtime/asm_amd64.s:1598 +0x1 fp=0xc00005cfe8 sp=0xc00005cfe0 pc=0x5cf621
created by runtime.init.6
        /home/omegarogue/go/go1.20.2/src/runtime/proc.go:293 +0x25

goroutine 3 [GC sweep wait]:
runtime.gopark(0x0?, 0x0?, 0x0?, 0x0?, 0x0?)
        /home/omegarogue/go/go1.20.2/src/runtime/proc.go:381 +0xd6 fp=0xc00005d780 sp=0xc00005d760 pc=0x5a00b6
runtime.goparkunlock(...)
        /home/omegarogue/go/go1.20.2/src/runtime/proc.go:387
runtime.bgsweep(0x0?)
        /home/omegarogue/go/go1.20.2/src/runtime/mgcsweep.go:278 +0x8e fp=0xc00005d7c8 sp=0xc00005d780 pc=0x58a58e
runtime.gcenable.func1()
        /home/omegarogue/go/go1.20.2/src/runtime/mgc.go:178 +0x26 fp=0xc00005d7e0 sp=0xc00005d7c8 pc=0x57f866
runtime.goexit()
        /home/omegarogue/go/go1.20.2/src/runtime/asm_amd64.s:1598 +0x1 fp=0xc00005d7e8 sp=0xc00005d7e0 pc=0x5cf621
created by runtime.gcenable
        /home/omegarogue/go/go1.20.2/src/runtime/mgc.go:178 +0x6b

goroutine 4 [GC scavenge wait]:
runtime.gopark(0xc00007c000?, 0x1925c48?, 0x1?, 0x0?, 0x0?)
        /home/omegarogue/go/go1.20.2/src/runtime/proc.go:381 +0xd6 fp=0xc00005df70 sp=0xc00005df50 pc=0x5a00b6
runtime.goparkunlock(...)
        /home/omegarogue/go/go1.20.2/src/runtime/proc.go:387
runtime.(*scavengerState).park(0x2909440)
        /home/omegarogue/go/go1.20.2/src/runtime/mgcscavenge.go:400 +0x53 fp=0xc00005dfa0 sp=0xc00005df70 pc=0x5884d3
runtime.bgscavenge(0x0?)
        /home/omegarogue/go/go1.20.2/src/runtime/mgcscavenge.go:628 +0x45 fp=0xc00005dfc8 sp=0xc00005dfa0 pc=0x588aa5
runtime.gcenable.func2()
        /home/omegarogue/go/go1.20.2/src/runtime/mgc.go:179 +0x26 fp=0xc00005dfe0 sp=0xc00005dfc8 pc=0x57f806
runtime.goexit()
        /home/omegarogue/go/go1.20.2/src/runtime/asm_amd64.s:1598 +0x1 fp=0xc00005dfe8 sp=0xc00005dfe0 pc=0x5cf621
created by runtime.gcenable
        /home/omegarogue/go/go1.20.2/src/runtime/mgc.go:179 +0xaa

goroutine 5 [finalizer wait]:
runtime.gopark(0x1a0?, 0x2909ae0?, 0x60?, 0x78?, 0xc00005c770?)
        /home/omegarogue/go/go1.20.2/src/runtime/proc.go:381 +0xd6 fp=0xc00005c628 sp=0xc00005c608 pc=0x5a00b6
runtime.runfinq()
        /home/omegarogue/go/go1.20.2/src/runtime/mfinal.go:193 +0x107 fp=0xc00005c7e0 sp=0xc00005c628 pc=0x57e827
runtime.goexit()
        /home/omegarogue/go/go1.20.2/src/runtime/asm_amd64.s:1598 +0x1 fp=0xc00005c7e8 sp=0xc00005c7e0 pc=0x5cf621
created by runtime.createfing
        /home/omegarogue/go/go1.20.2/src/runtime/mfinal.go:163 +0x45

rax    0x0
rbx    0x64283
rcx    0x7fe8b69a58ec
rdx    0x6
rdi    0x64283
rsi    0x64283
rbp    0x7fe8b5c08d40
rsp    0x7ffe6e31eef0
r8     0x4ad27d0
r9     0x0
r10    0x8
r11    0x246
r12    0x7fe8b793de28
r13    0x6
r14    0x7ffe6e31f000
r15    0x482d300
rip    0x7fe8b69a58ec
rflags 0x246
cs     0x33
fs     0x0
gs     0x0

Process finished with the exit code 2
```

**Comment by OmegaRogue** [comment]
Seems like that is because internally, Gtk actually checks if the input stream belongs to a file, but implementing my own MediaStream doesnt seem to work as the interface contains unexported methods. How do i do this?

**Comment by diamondburned** [comment]
Ah, I think I remember why I don't actually do this now. AFAIK, GTK never actually implemented video streaming on their code. You should double-check this with `gdb`, but I remember asking one of the GTK developers about this.

**Comment by OmegaRogue** [comment]
So, is the only way to do this with gstreamer and the gtk4 gstreamer plugin, is there another way to do it or does that also not work?

**Comment by diamondburned** [comment]
It would work if you download it as a file and play that and implement your own `GtkPaintable`, which would then allow you to use any video playback backend (incl. go-gst) to play it back.

My only concern is that subclassing is broken in gotk4 right now. I haven't touched that code in a bit, and I was still in the process of implementing it, so that unfortunately will likely not work.

Maybe `GtkPaintable` has something else. Maybe `GtkDrawingArea` could work? I'm not sure.

**Comment by OmegaRogue** [comment]
Hm, i dont think downloading it is really possible, as the video is live and should be played with minimal latency, this is essentially a remote desktop app, everything except for the video is working so far, and the video aswell if i output it to stdout and pipe my app into mpv

**Comment by diamondburned** [comment]
How many frames per second is the video? Is it a low FPS one like a CCTV camera or is it a 1080p60 live stream?

**Comment by OmegaRogue** [comment]
It's as high fps as possible, with arbitrary resolutions. This is a native linux client for the program [Weylus](https://github.com/H-M-H/Weylus), as all linux browsers I've tested barely support stylus input

**Comment by diamondburned** [comment]
Hmm, from reading the latest GTK source code, the Gst backend will play if you give it a URI somehow, so you can try and fiddle with that.

The FFmpeg backend should work for `InputStream`s. I wonder if there's a way to force use of FFmpeg.

**Comment by OmegaRogue** [comment]
Where are the docs of those backends? I haven't found those backends yet in my reading of the MediaStream and GtkVideo docs

**Comment by diamondburned** [comment]
I don't really know where the docs for those are. I'm just `grep`ping through the GTK source code right now.

Here's the line of code that's causing this issue: https://gitlab.gnome.org/GNOME/gtk/-/blob/main/modules/media/gtkgstmediafile.c#L299

**Comment by OmegaRogue** [comment]
I just asked about this on the GTK matrix rooms, and was pointed towards this https://gitlab.freedesktop.org/gstreamer/gst-plugins-rs/-/tree/main/video/gtk4 which provides a gdk.Paintable for the video

**Comment by diamondburned** [comment]
Aha! I found a hack to play back an HTTP video stream:

```go
package main

import (
	"fmt"
	"os"

	"github.com/diamondburned/gotk4/pkg/gio/v2"
	"github.com/diamondburned/gotk4/pkg/gtk/v4"
)

func main() {
	app := gtk.NewApplication("com.github.diamondburned.gotk4.giotest", 0)
	app.ConnectActivate(func() { activate(app) })
	app.Run(os.Args)
}

func activate(app *gtk.Application) {
	vfs := gio.VFSGetLocal()
	file := vfs.FileForURI("https://cdn.discordapp.com/attachments/342633910322069505/1087990964372647978/LLOGE.mp4")
	fmt.Println(file.URI())

	video := gtk.NewVideo()
	video.SetFile(file)

	win := gtk.NewApplicationWindow(app)
	win.SetChild(video)
	win.Show()
}
```

I will definitely be using this trick for my applications from now on.

**Comment by OmegaRogue** [comment]
That requires an http video tho, ill still have to use gstreamer for my app

**Comment by diamondburned** [comment]
I can't really tell exactly how the application works. I'd imagine whatever else you're using for your app will also work, as long as it has a valid and well-supported URI format for it. If not, you can always have a running FFmpeg process to transform it into a pipe.

**Comment by OmegaRogue** [comment]
I have a websocket on which i send input events and also send requests for frames in json, and then get Binary Messages back containing the frame encoded in h264, how would that work with ffmpeg? Gstreamer should probably work, but it's alot of setup and alot of additional complexity, as I haven't used Gstreamer before. When i get back a GtkPaintable from go-gst, can i just use type assertion to get the gotk4 type, or is there something else needed to get the proper type?
[go-gst returns property values as glib.Value from go-glib](https://pkg.go.dev/github.com/tinyzimmer/go-gst@v0.2.33/gst#ChildProxy.GetProperty), does that play well together with gotk4?

**Comment by diamondburned** [comment]
> [go-gst returns property values as glib.Value from go-glib](https://pkg.go.dev/github.com/tinyzimmer/go-gst@v0.2.33/gst#ChildProxy.GetProperty), does that play well together with gotk4?

No, you'll have to effectively "pass ownership" from go-glib to gotk4/glib, since they're completely different libraries.

You can also use gotk4 to autogenerate your own Gst bindings.

> I have a websocket on which i send input events and also send requests for frames in json, and then get Binary Messages back containing the frame encoded in h264, how would that work with ffmpeg?

I'm not sure if `GtkVideo` can play back an RTMP stream, but if it can, you can make FFmpeg start a local RTMP stream. Gst seems to support DASH and HLS.

> When i get back a GtkPaintable from go-gst, can i just use type assertion to get the gotk4 type, or is there something else needed to get the proper type?
> [go-gst returns property values as glib.Value from go-glib](https://pkg.go.dev/github.com/tinyzimmer/go-gst@v0.2.33/gst#ChildProxy.GetProperty), does that play well together with gotk4?

I'm actually unsure if go-gst can do this. go-gst doesn't have bindings to GTK, and Gst doesn't use GTK (it's the case the other way around), so how would it give you a `GtkPaintable`?

**Comment by OmegaRogue** [comment]
Seems like even with RTMP there is no way around writing a custom MediaStream or GFile implementation for it. How would I go about generating my own Gst bindings with gotk4?

With go-gst I thought of maybe just hoping that go lets me force the variable into the correct type somehow, but that probably wouldn't work

**Comment by diamondburned** [comment]
> Seems like even with RTMP there is no way around writing a custom MediaStream or GFile implementation for it.

You're not doing either. You'd probably be starting a local HTTP or TCP streaming server using one of those protocols and hope that Gst understands the URL. The FFmpeg implementation itself understands `InputStream`s, so the original code might work.

> With go-gst I thought of maybe just hoping that go lets me force the variable into the correct type somehow, but that probably wouldn't work

It wouldn't, no. Each library has to do its own memory management that effectively involves "keep some internal state with this and take its memory ownership"; using it in both libraries would very likely cause memory leaks.

**Comment by OmegaRogue** [comment]
![Peek 2023-03-22 23-26](https://user-images.githubusercontent.com/24958536/227052309-0160cd16-d60b-4779-af51-3ed8b9c48bad.gif)
Well, this is certainly progress, it doesnt quite work yet, but at least it doesnt crash

**Comment by OmegaRogue** [comment]
After more testing, it does not seem to be working or viable

**Comment by OmegaRogue** [comment]
[Here is the code in question](https://github.com/OmegaRogue/weylus-desktop/blob/67b21a9e72f4ea0cca9c3055af2086f8d0df0721/cmd/root.go#L106)

**Comment by OmegaRogue** [comment]
After alot of talk in the GTK matrix room, it definetly seems like the best way would be to generate bindings for gstreamer, how would i go about doing that?

**Comment by diamondburned** [comment]
Here's some code that reads a BMP from a tmpfs file: https://github.com/diamondburned/bmp-stream/blob/main/main.go. It achieves reasonable frame rates sometimes, but there's still some latency involved.

It might be worth writing a bit of C code to implement your own `GdkPaintable`. Ideally, you'd want these APIs for it:

```go
type FramePaintable struct { ptr *C.app_frame_paintable }

// Free frees the FramePaintable from memory. You would probably call unref() in here.
func (*FramePaintable) Free()

// Swap swaps the internal buffer with buf, returning the internal to the caller.
// This operation invalidates the Paintable.
func (*FramePaintable) Swap(buf *byte) *byte
```

Then, similar to the above code, you can swap the written buffer in, get the internal buffer out, and write to that the next time (double buffering).

**Reply by diamondburned** [reply]
It might also be worth it to do a CPU profile on the `bmp-stream` code to see what really needs optimization.

**Reply by diamondburned** [reply]
For completeness, here's bmp-stream implemented as a DrawingArea using Cairo instead of using the newer MemoryTexture API: https://github.com/diamondburned/bmp-stream/tree/cairo. This performs a bit worse on my machine, likely because it's drawn purely using software, despite having less allocations (I'm guessing).

**Reply by OmegaRogue** [reply]
So, after quite a while I finally got around to testing this, and it works, somewhat

https://user-images.githubusercontent.com/24958536/228929439-7598f0a9-26f8-4984-9fe8-d3becf91a114.mp4

The latency is quite high, but at least its better than not working, i used the code from https://github.com/diamondburned/bmp-stream/blob/main/main.go

**Reply by OmegaRogue** [reply]
With hardware decoding i've been able to get it a bit faster, closer to what your bmp-stream code achieves, but still alot more latency than wanted

**Reply by OmegaRogue** [reply]
A CPU Profile of bmp-stream seems to show that most of the time is spent on runtime.cgocall

**Comment by OmegaRogue** [comment]
Another option might be to write c code that uses gstreamer and the gtk4 gstreamer plugin and then returns the gdk paintable back to go and then just use that if thats possible, it would probably be alot simpler than trying to optimize streaming bitmaps to close to zero latency for arbitrary resolutions

**Reply by OmegaRogue** [reply]
How would i do that casting? I haven't quite been able to figure that out

**Reply by diamondburned** [reply]
Casting? If you can get the C code to return a `Paintable`, then you can just create an object like that:

```go
ptr := C.custom_get_paintable(&v.native) // function takes ref before returning
obj := coreglib.AssumeOwnership(unsafe.Pointer(ptr))
return obj.Cast().(gdk.Paintabler)
```

**Reply by OmegaRogue** [reply]
![image](https://user-images.githubusercontent.com/24958536/229212704-547a7159-a9d0-40f6-9984-d35a218799c2.png)
Ok, gstreamer-gtk4 seems to be working (the entire code thats running is c currently, but its being run through cgo), now i need to pass the paintable back to go and figure out a way to get the video stream into there

**Reply by OmegaRogue** [reply]
Now. in the latest commit of my project I added enough working go bindings for Gstreamer to get the same thing to run using gotk4

https://github.com/OmegaRogue/weylus-desktop/tree/main/gstreamer

**Comment by c1ngular** [comment]
nice job , @OmegaRogue , would you consider making your work as a independent package , or an additional subpackage to gotk4  @diamondburned , this would be a really useful feature to this project , i.e. a hardware accelerated media/streaming player , i have not seen any go UI toolkit/lib offers such a feature , most of them could not event deliver a simple video player .

**Reply by OmegaRogue** [reply]
ive already made it, tho it's not finished: https://github.com/OmegaRogue/gotk4-gstreamer

**Reply by OmegaRogue** [reply]
> ive already made it, tho it's not finished: https://github.com/OmegaRogue/gotk4-gstreamer

It works mostly iirc, but there are some dirty hacks in it that should be fixed

**Reply by c1ngular** [reply]
Great , a little guidelines how to use the package or examples would be perfect ,I would love to try this on ARM SBC (not received yet),  any tips on this ? @OmegaRogue  @diamondburned  .

Thanks

**Reply by OmegaRogue** [reply]
i had some uncommited local changes which seem to have been maybe important, its been a while since i had the time to work on this code, i pushed it to a new branch, including an example which i previously had local only
https://github.com/OmegaRogue/gotk4-gstreamer/blob/develop/pkg/_examples/basic/main.go

**Reply by OmegaRogue** [reply]
I'm not quite sure if the changes added in develop are necessary for that version of the example to work

**Reply by c1ngular** [reply]
Thanks again, I will give a try when I got my arm-based SBC , my goal is to make an application that plays/views multiple rtsp camera feeds simultaneously in same window, I will get back here with my findings .

**Reply by diamondburned** [reply]
> ive already made it, tho it's not finished: https://github.com/OmegaRogue/gotk4-gstreamer

This is awesome!

> including an example which i previously had local only https://github.com/OmegaRogue/gotk4-gstreamer/blob/develop/pkg/_examples/basic/main.go

Just a small note, `func produceImageFrame` here is unused.

**Reply by c1ngular** [reply]
@OmegaRogue  i could not get example running [gotk4-gstreamer](https://github.com/OmegaRogue/gotk4-gstreamer) from both main and develop branches (on Apple M1 Pro) , it complains a lot about type mismatch in assignment and conversion [issue](https://github.com/OmegaRogue/gotk4-gstreamer/issues/2#issue-1823526146) , though i can successfully run gotk4 examples , did i miss something ? 

thanks

**Reply by diamondburned** [reply]
This sounds like some difference in architecture/platform. Perhaps you might want to re-run `go generate` on the repository.

**Reply by c1ngular** [reply]
`go generate ` shows error :

go: downloading github.com/kr/pretty v0.3.1
go: downloading github.com/kr/text v0.2.0
go: downloading github.com/rogpeppe/go-internal v1.9.0
2023/08/04 09:11:24 error adding package "gstreamer-app-1.0": only 10 girs found
exit status 1

**Reply by diamondburned** [reply]
Are you already in the Nix shell? Try `nix-shell` in the repo.

**Reply by OmegaRogue** [reply]
I think know that issue, there should be code to manually fix it after generation in there, maybe its commented out
